### PR TITLE
[17.0][MIG] edi_oca: Migration to 17.0

### PR DESCRIPTION
**Pull request by john-herholz-dt**
A lot of views needed to be adjusted.

Furthermore I changed the _search method on edi_exchange_record, as the count param no longer works.
Not sure about the change as I don't know what the override of the _search method was suppossed to do.

Edit: @nguyenminhchien helped on the _search method.

### CODE REVIEW

File: edi_oca/models/edi_exchange_record.py [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+        self._trigger_edi_event("done", suffix="ack_missing")
+
+    def _notify_ack_received_error(self):
+        self._notify_related_record(
+            self._exchange_status_message("ack_received_error"),
+        )
+        self._trigger_edi_event("done", suffix="ack_received_error")
+
+    @api.model
+    def _search(
```
Comment (john-herholz-dt) [inline edi_oca/models/edi_exchange_record.py]: Someone have a look at the changes here, please.

File: edi_oca/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# @author: Simone Orsi <simahawk@gmail.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    "name": "EDI",
+    "summary": """
+    Define backends, exchange types, exchange records,
+    basic automation and views for handling EDI exchanges.
+    """,
+    "version": "17.0.1.1.0",
```
Comment (nguyenminhchien) [inline edi_oca/__manifest__.py]: it should be `17.0.1.0.0`

File: edi_oca/migrations/16.0.1.0.0/pre-migrate.py
Comment (nguyenminhchien) [inline edi_oca/migrations/16.0.1.0.0/pre-migrate.py:1]: please remove the `migrations` folder

File: edi_oca/models/edi_exchange_record.py [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+        """Handy method to not have to convert b64 back and forth."""
+        self.ensure_one()
+        if not self[field_name]:
+            return ""
+        if binary:
+            res = base64.b64decode(self[field_name])
+            return res.decode() if not as_bytes else res
+        return self[field_name]
+
+    def name_get(self):
```
Comment (nguyenminhchien) [inline edi_oca/models/edi_exchange_record.py]: you should now override `_compute_display_name` instead, adding the possible new fields in the depends of the method. See https://github.com/odoo/odoo/pull/122085 for more details.
For reference:
```
    @api.depends("identifier", "res_id", "model", "type_id")
    def _compute_display_name(self):
        for rec in self:
            rec_name = rec.identifier
            if rec.res_id and rec.model:
                rec_name = rec.record.display_name
            rec.display_name = f"[{rec.type_id.name}] {rec_name}"

```
Seems that you miss this document: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0#tasks-to-do-in-the-migration 
please follow its guideline

File: edi_oca/models/edi_exchange_record.py (hunk 1 above)
Comment (nguyenminhchien) [reply edi_oca/models/edi_exchange_record.py]: I have a suggestion here, please try it:

```

    @api.model
    def _search(self, domain, offset=0, limit=None, order=None, access_rights_uid=None):
        query = super()._search(
            domain,
            offset=offset,
            limit=limit,
            order=order,
            access_rights_uid=access_rights_uid,
        )

        if self.env.is_system():
            # restrictions do not apply to group "Settings"
            return query

        # TODO highlight orphaned EDI records in UI:
        #  - self.model + self.res_id are set
        #  - self.record returns empty recordset
        # Remark: self.record is @property, not field

        if query.is_empty():
            return query
        orig_ids = list(query)
        ids = set(orig_ids)
        result = []
        model_data = defaultdict(
            lambda: defaultdict(set)
        )  # {res_model: {res_id: set(ids)}}
        for sub_ids in self._cr.split_for_in_conditions(ids):
            self._cr.execute(
                """
                            SELECT id, res_id, model
                            FROM "%s"
                            WHERE id = ANY (%%(ids)s)"""
                % self._table,
                dict(ids=list(sub_ids)),
            )
            for eid, res_id, model in self._cr.fetchall():
                if not model:
                    result.append(eid)
                    continue
                model_data[model][res_id].add(eid)

        for model, targets in model_data.items():
            if not self.env[model].check_access_rights("read", False):
                continue
            recs = self.env[model].browse(list(targets))
            missing = recs - recs.exists()
            if missing:
                for res_id in missing.ids:
                    _logger.warning(
                        "Deleted record %s,%s is referenced by edi.exchange.record %s",
                        model,
                        res_id,
                        list(targets[res_id]),
                    )
                recs = recs - missing
            allowed = (
                self.env[model]
                .with_context(active_test=False)
                ._search([("id", "in", recs.ids)])
            )
            for target_id in allowed:
                result += list(targets[target_id])
        if len(orig_ids) == limit and len(result) < len(orig_ids):
            extend_query = self._search(
                domain,
                offset=offset + len(orig_ids),
                limit=limit,
                order=order,
                access_rights_uid=access_rights_uid,
            )
            extend_ids = list(extend_query)
            result.extend(extend_ids[: limit - len(result)])
        # Restore original ordering
        result = [x for x in orig_ids if x in result]
        if set(orig_ids) != set(result):
            # Create a virgin query
            query = self.browse(result)._as_query()
        return query
```

File: edi_oca/static/src/xml/widget_edi.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates id="template" xml:space="preserve">
+    <div t-name="edi_oca.EdiConfigurationWidget" owl="1">
```
Comment (nguyenminhchien) [inline edi_oca/static/src/xml/widget_edi.xml]: please remove `owl="1"`

File: edi_oca/views/edi_exchange_type_rule_views.xml [hunk 5]
```diff
@@ -0,0 +1,124 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <record id="edi_exchange_type_rule_view_tree" model="ir.ui.view">
+        <field name="model">edi.exchange.type.rule</field>
+        <field name="arch" type="xml">
+            <tree decoration-muted="(not active)">
+                <field name="active" invisible="1" />
```
Comment (nguyenminhchien) [inline edi_oca/views/edi_exchange_type_rule_views.xml]:
```suggestion
                <field name="active" column_invisible="1" />
```

File: edi_oca/views/edi_exchange_type_views.xml [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                            name="model_rules"
+                            string="Model rules"
+                            groups="edi_oca.group_edi_advanced_settings_manager"
+                        >
+                            <field
+                                name="rule_ids"
+                                context="{'default_type_id': active_id}"
+                            >
+                                <tree decoration-muted="(not active)">
+                                    <field name="active" invisible="1" />
```
Comment (nguyenminhchien) [inline edi_oca/views/edi_exchange_type_views.xml:75]:
```suggestion
                                    <field name="active" column_invisible="1" />
```

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]
PR of `base_edi` has been merged, you can remove it from `test-requirements.txt`

**Comment by john-herholz-dt** [comment]
Thanks @nguyenminhchien for your in depth review. Will adjust everything later today.
Sure I might have forgotten about some essential changes like the name_get. I didn't look into it.

File: edi_oca/models/edi_exchange_record.py (hunk 1 above)
Comment (john-herholz-dt) [reply edi_oca/models/edi_exchange_record.py]: Thanks for the suggestion!

**Comment by john-herholz-dt** [comment]
I have applied the suggested changes, all tests now pass.

**Review by nguyenminhchien — APPROVED** [review]

**Comment by john-herholz-dt** [comment]
Hello @simahawk , I know you are busy, but once you have some minutes, could you check this PR, I need to migrate everything to 17.0 and edi_oca is the most critical.

I also saw there was another PR for 16 "add partner form page". Do I need to update the code?
Does this work with the git patch command from the migration wiki?

Thank you very much!

**Comment by john-herholz-dt** [comment]
I have re-done the migration with the latest changes from 16.0 branch that were not included previously.

**Comment by john-herholz-dt** [comment]
@simahawk what are we doing with this?

**Review by JordiMForgeFlow — APPROVED** [review]
Code review

**Comment by peluko00** [comment]
Can you rebase please for the runboat @john-herholz-dt ?

**Comment by john-herholz-dt** [comment]
I did the rebase, somehow it had now problems with the pre-commit checks, I did another one pre-commit on my side

**Comment by peluko00** [comment]
the last commit will be include with the migration commit

> I did the rebase, somehow it had now problems with the pre-commit checks, I did another one pre-commit on my side

**Review by peluko00 — APPROVED** [review]
Code review, LGTM

**Comment by john-herholz-dt** [comment]
Finally! Jeehaaaa.

**Pull request merged**